idx/memory/tag_query.go

```diff
@@ -0,0 +1,176 @@
+package memory
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/raintank/metrictank/idx"
+)
+
+var (
+	expressionRe    = regexp.MustCompile("^([^!=~]+)([!=~]{1,3})(.*)$")
+	errUnknownId    = errors.New("id does not exist")
+	errInvalidTag   = errors.New("found an invalid tag in the index")
+	errInvalidQuery = errors.New("invalid query")
+)
+
+type TagQuery struct {
+	selects []func(TagIndex) (map[string]struct{}, error)
+	filters []func(def *idx.Archive) (bool, error)
+}
+
+func NewTagQuery(expressions []string) (*TagQuery, error) {
+	if len(expressions) == 0 {
+		return nil, errInvalidQuery
+	}
+
+	query := &TagQuery{}
+
+	for _, expr := range expressions {
+		if !expressionRe.MatchString(expr) {
+			return nil, errInvalidQuery
+		}
+
+		exprSplits := expressionRe.FindAllStringSubmatch(expr, -1)
+		if len(exprSplits[0]) != 4 {
+			return nil, errInvalidQuery
+		}
+
+		key := exprSplits[0][1]
+		operator := exprSplits[0][2]
+		value := exprSplits[0][3]
+
+		// special case of empty value
+		if len(value) == 0 {
+			if operator[0] == byte('=') {
+				operator = "!=~"
+				value = ".+"
+			} else {
+				operator = "=~"
+				value = ".+"
+			}
+		}
+
+		// always anchor all regular expressions at the beginning
+		if operator[len(operator)-1] == byte('~') && len(value) > 0 && value[0] != byte('^') {
+			value = fmt.Sprintf("^%s", value)
+		}
+
+		switch operator {
+		case "=":
+			query.selects = append(query.selects, func(idx TagIndex) (map[string]struct{}, error) {
+				return idx[key][value], nil
```

Comment:
can `idx[key]` sometimes be nil?

Reply:
yes, but that's ok as long as we only read and don't write

Reply:
but looking up the key `value` in a nil map will generate a panic.

Reply:
I don't think that's true: https://play.golang.org/p/QvFBden7pD

Reply:
whoa! very interesting.